### ISSUE TITLE
Row styling

### ISSUE THREAD

**Issue by codener** [matt-harvey/tabulo#9]
Dear Matt,

thanks for this awesome library. I just replaced a bunch of half-working custom table-printing code with Tabulo, and it's great.

There is one thing I could not figure out, however: how can I style a whole row depending on the row record?

My example is printing a list of hashes as a table. Each hash contains a `:color` key (for which I did not add a column). I would like to color the whole row in the specified color. How can I achieve this?

**Comment by matt-harvey** [comment]
Thanks for the feedback @codener . Great to here that you've found Tabulo useful!

As for your question, colors are applied using the `styler` param that can be passed either to `Tabulo::Table.new` or to `Tabulo::Table#add_column`. This param accepts a callback, which has access both to the underlying value of each cell, and also to each string fragment (one per wrapped line) within the cell.

Unfortunately, there is no straightforward way of doing what you ask, since the `styler` callback does not have access either to the member of the original underlying data set, or the `Tabulo::Row` instance for its row. It can only access what's in its particular cell.

To work around this, you can transform the original collection by incorporating color information into each individual cell. Here's one way of doing this:

```ruby
require "tabulo"
require "rainbow"

original_data = [
  { color: :red, name: "Alice", height: 160 },
  { color: :blue, name: "Bob", height: 170 },
  { color: :green, name: "Charles", height: 180 },
]

# Convert the data into something like:
# [
#   { name: { color: :red, content: "Alice" }, height: { color: :red, content: 160 } },
#   { name: { color: :blue, content: "Bob" }, height: { color: :blue, content: 170 } },
#   { name: { color: :green, content: "Charles" }, height: { color: :green, content: 180 } },
# ]
data_with_colors = original_data.map do |datum|
  color = datum.delete(:color)
  datum.map { |key, value| [key, { color: color, content: value }] }.to_h
end

styler = -> (cell_value, str) { Rainbow(str).public_send(cell_value[:color]) }
formatter = -> (cell_value) { cell_value[:content].to_s }

table = Tabulo::Table.new(data_with_colors, formatter: formatter, styler: styler) do |t|
  t.add_column("Name", styler: styler) { |s| s[:name] }
  t.add_column("Height", styler: styler, align_body: :right) { |s| s[:height] }
end

puts table.pack
```

It's a bit ungainly I admit.

Another workaround is to apply colors to each row as a whole, during the actual printing of the table, something like this:

```ruby
require "tabulo"
require "rainbow"

data = [
  { color: :red, name: "Alice", height: 160 },
  { color: :blue, name: "Bob", height: 170 },
  { color: :green, name: "Charles", height: 180 },
]

table = Tabulo::Table.new(data, border: :blank) do |t|
  t.add_column("Name") { |s| s[:name] }
  t.add_column("Height") { |s| s[:height] }
end

table.zip(data).each do |row, datum|
  puts Rainbow(row).public_send(datum[:color])
end
```

This needs some additional tweaks, however, to avoid applying colors to the header row, which is rendered as part of the first table row. (The tweak would involve splitting by the newline character etc..)

Both these workarounds are a little awkward I admit. To make it less awkward the `styler` callback would have to have an additional parameter providing access to the underlying enumerable member. I'm not sure that I want to add that to the API at this point; I don't want to break backwards compatibility or to complicate the API too much. (And Tabulo is not ready for a version 3 release just yet.)

Let me know if either of the above workarounds is useful for you.

**Comment by codener** [comment]
Thanks for the examples. I believe the first is in the spirit of Tabulo and the best solution currently.

However, adding access to the underlying record as you suggest would be a reasonable addition in my eyes. To make it a  backwards-compatible change, you could check the `arity` of the `styler` lambda (and maybe the `formatter` as well) and pass just as many arguments as the lambda expects.

My use case was not to get a quick overview of record data, or to print some Excel sheet to the console, but to lay out structured data without wrapping, manual padding etc. Being able to reference the row data from a cell would add valuable flexibility.

As for my original issue, I went with coloring only a single cell and moved the color-deciding logic to the styler of that cell. Good enough in that case.

**Comment by matt-harvey** [comment]
Thanks @codener, this is valuable feedback.

I'm glad you were able to find a good-enough solution to your original issue.

Yes you're right, checking the `arity` is an option to preserve backwards compatibility. I will give this serious consideration. I will leave this issue open for now, and will return to it when I have the time to properly work through the options and trade-offs. I want to think carefully about how I evolve the API from here. In the meantime, further suggestions are welcome.

**Comment by codener** [comment]
Let me know if I can be of any help

Am So., 15. März 2020 um 23:16 Uhr schrieb Matt Harvey <
notifications@github.com>:

**Comment by matt-harvey** [comment]
Thanks @codener 

I've mulled this over, and I think the best solution is to add an optional second parameter to the `formatter` callback, and an optional third parameter to the `styler` callback. In both cases, the extra parameter would be a struct instance. The struct would be called something like `CellData` and would have an attribute, `source`, that returns the original row record (called `source` to match the `sources` parameter that is passed to the `Table` initializer). It could also expose a `row_index` and `column_index` attribute, which would facilitate, for example, coloring alternate rows or alternate colors differently.

By making the additional parameter a struct, it's extensible with further attributes if required in future, without proliferating positional callback parameters or breaking backwards compatibility.

A second solution that I've considered is to wait until a v3 release and then replace the second parameter of `styler` with a `CellData` parameter, and have `CellData` also contain the underlying cell value as an attribute. (For context: I am considering a v3 release for late 2020, as there are some other changes I want to make that I am saving for a major version upgrade.) This would avoid the additional complexity inherent in having more than one variant of the `styler` callback.

I think I like the first solution better though: having the `CellData` parameter be optional in _both_ the `formatter` _and_ `styler` callbacks provides a degree of consistency between those two. (For ergonomic reasons, I want the `formatter` callback to have at least one variant that takes a single parameter being the cell value, and therefore I don't want `CellData` to be a compulsory parameter for that callback.) Further, if the `CellData` parameter is optional, there's a small optimization benefit, since there's no need a initialize a `CellData` instance on every call unless we know that the callback requires it.

Thoughts?

**Comment by codener** [comment]
Solution no. 1 sounds reasonable and flexible. Thanks for mulling :)

**Comment by matt-harvey** [comment]
No probs :) Thanks for your feedback.

I'll work on this over the weekend. Happy to implement it all myself; but if you are keen to contribute a PR, let me know and we can coordinate.

**Comment by matt-harvey** [comment]
The proposed enhancement to the `styler` callback has now been implemented on the [v2-4](https://github.com/matt-harvey/tabulo/tree/v2-4) branch.

I'll make similar enhancements to the `formatter` and `header_styler` callbacks, and possibly also the `extractor` callback. Once these are done I'll release v2.4.0. This will likely be some time in the next few days. Please regard the [v2-4](https://github.com/matt-harvey/tabulo/tree/v2-4) branch as unstable in the meantime.

**Comment by matt-harvey** [comment]
Enhancement to `formatter` also implemented on [v2-4](https://github.com/matt-harvey/tabulo/tree/v2-4) branch.

**Comment by matt-harvey** [comment]
Enhancement to `header_styler` now also implemented on [v2-4](https://github.com/matt-harvey/tabulo/tree/v2-4) branch.

**Comment by matt-harvey** [comment]
Enhancement to `extractor` callback implemented on [v2-4](https://github.com/matt-harvey/tabulo/tree/v2-4) branch.

Changes merged into `master` branch and **v2.4.0** released.

**Comment by codener** [comment]
Thank you very much!
